### ISSUE TITLE
Add text direction property in the Abstract Code Editor

### ISSUE THREAD

**Issue by blink1073** [jupyterlab/jupyterlab#1359]
This depends on #1140.

cf https://github.com/jupyterlab/jupyterlab/pull/1140#issuecomment-261765179.

After that, open a new issue to track the implementation of the text direction.

**Comment by blink1073** [comment]
cc @tomerm

**Comment by tomerm** [comment]
The idea is to provide default text direction for content being authored in specific instance of rich text editor. This is a similar parameter to what is used in CKEditor: **contentsLangDirection** ( http://docs.ckeditor.com/#!/api/CKEDITOR.config-cfg-contentsLangDirection ). 
Possible values should be: 

1. ltr (left-to-right), 
2. rtl (right-to-left), 
3. contextual (or auto as used in HTML)

In addition to that, there should be explicit way for the end user to interactively change text direction for selected text (or for current paragraph in which cursor is positioned in case current selection is of zero length). This can be achieved via:

1. GUI buttons - similar to all rich text editors (i.e. http://ckeditor.com/addon/bidi)
AND / **OR** (in case there is no toolbars for any new buttons)
2. Keyboard shortcuts (i.e. in Notepad it is <Ctrl><Left|Right-Shift>)

CC @ashensis FYI.

**Comment by akosyakov** [comment]
It would be nice to report such requirements to editors of use, e.g. codemirror and monaco, that JLab can get away by delegating to them

**Comment by tomerm** [comment]
@akosyakov,  definitely, makes a lot of sense to me. 

Since we would also like to contribute a solution to codemirror / monaco (not just communicate the requirements) it would help a lot to understand if JLab is going with monaco already in the first release or at least for now it will be still based on codemirror. The reason is very simple: contributing 2 solutions (to codemirror **and** monaco) is significantly more expensive than doing just one. Thus I should be careful with decisions based on:

- the timing / plans for JLab and 
- actual resources funding available to me internally.

**Comment by akosyakov** [comment]
@tomerm there is already a similar issue in the backlog of vscode/monaco: https://github.com/Microsoft/vscode/issues/4994

**Comment by blink1073** [comment]
It sounds like we should wait to add this property to the interface until at least one of the used editors supports it.

**Comment by blink1073** [comment]
cf https://www.w3.org/International/tutorials/bidi-xhtml/

**Comment by blink1073** [comment]
cf https://www.unicodeconference.org/presentations-43/S5T3-Schottlender.pdf
